### DISCUSSION TITLE
Randomly returning false on flag check with Java SDK.

### DISCUSSION THREAD

**Discussion by ShurikAg** [Unleash/unleash#4318] [Q&A]
We have recently implemented feature flagging in our app (based on self-hosted Unleashed).

React side works perfectly as expected. However, the server-side (written in grails) is giving wrong results every now and then. It is inconsistent. I am not even sure where to start debugging it, but I am sure it is something in our code.

On another note: on local everything works as expected :( 

**Here is our setup in Unleash**

API Keys:
<img width="1223" alt="image" src="https://github.com/Unleash/unleash/assets/916432/b9a6e3f1-cd13-4f1c-8555-776c93231606">

One of the feature keys setup:
<img width="932" alt="image" src="https://github.com/Unleash/unleash/assets/916432/dd7b5a57-133b-4673-9947-380e8e66a180">

**Related code on the backend**
Controller wrapper:
```
trait FeaturesAwareController {

    UnleashService unleashService

    def isFeatureEnabled(UnleashFeature feature, User user, Closure closure) {
        if (unleashService.isFeatureEnabled(feature, user)) {
            closure()
        } else {
            throw new FeatureException("Feature not found")
        }
    }
}
```

Unleash service:
```
class UnleashService {

    UnleashClientProviderService unleashClientProviderService
    UnleashContextProviderService unleashContextProviderService

    Boolean isFeatureEnabled(UnleashFeature feature, User user) {
        unleashClientProviderService.getClient()
                .isEnabled(feature.name, unleashContextProviderService.getContext(user))
    }
}
```

UnleashContextProviderService:
```
class UnleashContextProviderService {

    UnleashContext getContext(User user) {
        UnleashContext.builder()
                    .userId(user.email)
                    .build()
    }
}
```

UnleashClientProviderService:
```
class UnleashClientProviderService {

    Unleash client

    @Value('${priz.unleash.appName}')
    String appName

    @Value('${priz.unleash.unleashAPI}')
    String unleashAPI

    @Value('${priz.unleash.apiKey}')
    String apiKey

    Unleash getClient() {
        if (!client) {
            UnleashConfig config = UnleashConfig.builder()
                    .appName(appName)
                    .unleashAPI(unleashAPI)
                    .apiKey(apiKey)
                    .build()

            client = new DefaultUnleash(config)
        }

        client
    }
}
```

Basically, what happens is it always returns the correct result on Fronend (in React). However, in most cases, the code above returns false, even if the feature is enabled.

The code is super simple and follows the docs. Maybe something in the configs in Unleash.

I am totally lost.

**Comment by gardleopard** [comment]
Hi @ShurikAg,

Sorry for the long waiting time. Our Java SDK should not work inconsistently. I suspect that the userId is not allways set when the Java SDK evaluates the toggle. Is it possible for you to log that userId to test if that is the issue?

**Reply by thomasheartman** [reply]
@ShurikAg Just checking in to see if you've had the time to find this response? If we don't hear from you in another week and a bit, we'll go ahead and close this for now.

**Reply by ShurikAg** [reply]
Hi @thomasheartman , I am trying to reproduce it again now, and looks like everything just started working... Nothing changed though, so I am not sure what is going on. The only difference is that I added logs for feature check.

I'll monitor that in the next few days and keep you posted.

**Reply by thomasheartman** [reply]
Huh, that's strange (...but also good?). But yes, please do keep us posted 🙏🏼

**Comment by daveleek** [comment]
Hey @ShurikAg

I'll close this for now as it seems things have worked out for you. But if you have more questions feel free to open a new issue!
